### PR TITLE
Implement no-attribute dispatch of ATen ops from the JIT

### PR DESCRIPTION
**Pull request by apaszke**
This will be needed to e.g. dynamically specify scalar arguments in JIT script.

### CODE REVIEW

File: tools/jit/gen_jit_dispatch.py [hunk 1]
```diff
@@ -58,64 +63,126 @@ def is_jit_op(decl):
...
+
+
 def gen_jit_dispatch(declarations, out):
     aten_decls = load_aten_declarations(declarations)
     jit_decls = [d for d in aten_decls if is_jit_op(d)]
 
     def is_tensor_arg(arg):
         return arg['simple_type'] in {'Tensor', 'TensorList'}
 
+    for decl in jit_decls:
```
Comment (zdevito) [inline tools/jit/gen_jit_dispatch.py]: This for doesn't do anything, I think...

File: tools/jit/templates/aten_dispatch.cpp [hunk 2]
```diff
@@ -142,6 +142,113 @@ std::array<bool, N> as_bool_array(const std::vector<int64_t>& vec) {
...
+T tensor_as(const at::Tensor& t) {
+  throw std::runtime_error("Invalid tensor_as overload");
+}
+
+template<>
+int64_t tensor_as(const at::Tensor& t) {
+  if (t.numel() > 1)
+    throw std::runtime_error("Trying to convert a tensor with more than 1 element into a scalar");
+  // TODO: transfer to CPU
+  switch (t.type().scalarType()) {
```
Comment (zdevito) [inline tools/jit/templates/aten_dispatch.cpp]: We can replace these with at::Scalar(t), the control flow PR shows how to do it.

File: tools/jit/gen_jit_dispatch.py [hunk 3]
```diff
@@ -58,64 +63,126 @@ def is_jit_op(decl):
...
+            attr_names = sorted([arg['name'] for arg in kw_scalar_args])
+            num_inputs = len(tensor_arg_idx) + len(pos_scalar_arg_idx) if not has_tensorlist else '*'
+            descriptor = '-'.join([decl['name'], str(num_inputs)] + attr_names)
+
+            kw_assignments = [KW_ASSIGNMENT.substitute(type_cast=TYPE_CASTS.get(arg['simple_type'], arg['simple_type']),
+                                                       name=arg['name'],
+                                                       method=ATTR_METHOD_MAP[arg['simple_type']])
+                              for arg in kw_scalar_args]
+            # TODO: if has_tensor_list then we need to count form the end!
+            pos_scalar_arg_offsets = [i + len([None for t_idx in tensor_arg_idx if t_idx < s_idx])
```
Comment (zdevito) [inline tools/jit/gen_jit_dispatch.py]: `[None for t_idx in tensor_arg_idx if t_idx < s_idx]` <--  This heavy use of array indexing is confusing. I can't really follow what is going on here. Can you find a clearer way of categorizing each input? I think a loop over all arguments, classifying them as tensor, position_scalar, or kwarg and adding entries into pos_assigns, kw_assignments, etc. would be easier to follow.

File: tools/jit/gen_jit_dispatch.py [hunk 4]
```diff
@@ -58,64 +63,126 @@ def is_jit_op(decl):
...
-        ops[descriptor] = constructor
+                # There are still some situations where we can recover from an ambiguous descriptor.
+                def fail():
+                    raise RuntimeError('descriptor conflict: ' + descriptor)
+
+                # If there are two overloads with the same descriptor, that differ only by a type of a
+                # single argument, where one of them takes a tensor, while another one takes an
+                # at::Scalar as a positional scalar arg, then prefer the tensor overload.
+                # It should get broadcasted correctly.
+                if descriptor in prefer_tensor_overload:
```
Comment (zdevito) [inline tools/jit/gen_jit_dispatch.py]: Is it possible to just have a list of exclusions instead? This recover code is pretty long. e.g. just have `[ ('lt-2', 1), ... ` and then have a check if the descriptor is `lt-2` and arg_type[1] == 'Scalar', then don't bother adding it.

**Review by zdevito — COMMENTED** [review]
This looks pretty good. I'd like to see a simpler way of handling the indexing computations for position scalar args, but other than that it is ready to go.

**Comment by apaszke** [comment]
Please re-review carefully. I had to change a few things because:
- there was a lot of conflicts
- computing stack offsets is a bit more annoying than indexing a list
- I added support for functions taking TensorList (it was missing previously)

**Review by zdevito — APPROVED** [review]
Looks good! Thanks for rebasing.

**Pull request merged**